pkg/trace/api/otlp.go

```diff
@@ -0,0 +1,506 @@
+		case *otlppb.AnyValue_DoubleValue:
+			span.Metrics[kv.Key] = v.DoubleValue
+		case *otlppb.AnyValue_IntValue:
+			span.Metrics[kv.Key] = float64(v.IntValue)
+		default:
+			span.Meta[kv.Key] = anyValueString(kv.Value)
+		}
+	}
+	if _, ok := span.Meta["env"]; !ok {
+		if env := span.Meta["deployment.environment"]; env != "" {
+			span.Meta["env"] = env
+		}
+	}
+	if in.TraceState != "" {
+		span.Meta["trace_state"] = in.TraceState
+	}
+	if lib.Name != "" {
+		span.Meta["instrumentation_library.name"] = lib.Name
+	}
+	if lib.Version != "" {
+		span.Meta["instrumentation_library.version"] = lib.Version
+	}
+	if svc := span.Meta["peer.service"]; svc != "" {
+		span.Service = svc
+	}
+	if r := resourceFromTags(span.Meta); r != "" {
+		span.Resource = r
+	}
+	span.Type = spanKind2Type(in.Kind, span)
+	status2Error(in.Status, in.Events, span)
+	return span
+}
+
+// resourceFromTags attempts to deduce a more accurate span resource from the given list of tags meta.
+// If this is not possible, it returns an empty string.
+func resourceFromTags(meta map[string]string) string {
+	var r string
+	if m := meta["http.method"]; m != "" {
+		r = m
+		if route := meta["http.route"]; route != "" {
+			r += " " + route
+		} else if route := meta["grpc.path"]; route != "" {
+			r += " " + route
+		}
+	} else if m := meta["messaging.operation"]; m != "" {
+		r = m
+		if dest := meta["messaging.destination"]; dest != "" {
+			r += " " + dest
+		}
+	}
+	return r
+}
+
+// status2Error checks the given status and events and applies any potential error and messages
+// to the given span attributes.
+func status2Error(status *otlppb.Status, events []*otlppb.Span_Event, span *pb.Span) {
+	if status == nil || status.Code != otlppb.Status_STATUS_CODE_ERROR {
+		return
+	}
+	span.Error = 1
+	for _, e := range events {
+		if strings.ToLower(e.Name) != "exception" {
+			continue
+		}
+		for _, attr := range e.Attributes {
+			switch attr.Key {
+			case "exception.message":
+				span.Meta["error.msg"] = anyValueString(attr.Value)
+			case "exception.type":
+				span.Meta["error.type"] = anyValueString(attr.Value)
+			case "exception.stacktrace":
+				span.Meta["error.stack"] = anyValueString(attr.Value)
+			}
+		}
+	}
+}
+
+// spanKind2Type returns a span's type based on the given kind and other present properties.
+func spanKind2Type(kind otlppb.Span_SpanKind, span *pb.Span) string {
+	var typ string
+	switch kind {
+	case otlppb.Span_SPAN_KIND_SERVER:
+		typ = "web"
+	case otlppb.Span_SPAN_KIND_CLIENT:
+		typ = "http"
+		db, ok := span.Meta["db.system"]
+		if !ok {
+			break
+		}
+		switch db {
+		case "redis", "memcached":
+			typ = "cache"
+		default:
+			typ = "db"
+		}
+	default:
+		typ = "custom"
+	}
+	return typ
+}
+
+func byteArrayToUint64(b []byte) uint64 {
+	if len(b) < 8 {
+		return 0
+	}
+	return binary.BigEndian.Uint64(b[len(b)-8:])
+}
+
+// anyValueString converts otlppb.AnyValue a to its string representation.
+func anyValueString(a *otlppb.AnyValue) string {
```

Comment:
On the metrics side we use https://github.com/open-telemetry/opentelemetry-collector/blob/3f5c7180c51ed67a6f54158ede5e523822e9659e/translator/trace/protospan_translation.go#L61 to make this translation and optionally add them as metric tags (the types are different because of the `pdata` package but it's the same thing). If this affects discoverability we should try and do the same translation.

Reply:
Not entirely sure what you're proposing, but I feel its best we keep this in our own hands because its on the hot path (runs on each span multiple times) and we want to get this as performant as possible. Furthermore the code you linked to can be quite slow if it has to marshal JSON using the `reflect` package...

Reply:
So, my concern is if the Datadog webapp uses these span metadata fields to correlate with Datadog metric tags.

Imagine we have a field `key` with value `0.9`. With your code this gets mapped to `"0.90"`, while with the OpenTelemetry one it gets mapped to `"0.9"` (see https://play.golang.org/p/O4q9xHQcguU). That means we would get a `key:0.9` tag on metrics and a `key:0.90` field on the trace metadata. If the Datadog webapp uses this to correlate the metric and the trace (not sure if it does) then it won't find one from the other since they have different strings as values. The same thing happens with other fields.

Reply:
I would leave it for now if it's ok with you, because this is definitely hot path and gets called a lot. The linked implementation uses `json.Marshal` which is very slow, while this one is optimised. If we run into issues, we can always change it (experimental).